edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Diagnostics/Metric.cs

```diff
@@ -3,45 +3,68 @@
 namespace Microsoft.Azure.Devices.Edge.Agent.Diagnostics
 {
     using System;
+    using System.Collections.Generic;
+    using System.Linq;
     using Microsoft.Azure.Devices.Edge.Util;
 
     public sealed class Metric : IEquatable<Metric>
     {
         public DateTime TimeGeneratedUtc { get; }
         public string Name { get; }
         public double Value { get; }
-        public string Tags { get; }
+        public IReadOnlyDictionary<string, string> Tags { get; }
 
-        public Metric(DateTime timeGeneratedUtc, string name, double value, string tags)
+        public readonly Lazy<int> MetricKey;
+
+        public Metric(DateTime timeGeneratedUtc, string name, double value, IReadOnlyDictionary<string, string> tags)
         {
             Preconditions.CheckArgument(timeGeneratedUtc.Kind == DateTimeKind.Utc, $"Metric {nameof(timeGeneratedUtc)} parameter only supports in UTC.");
             this.TimeGeneratedUtc = timeGeneratedUtc;
             this.Name = Preconditions.CheckNotNull(name, nameof(name));
             this.Value = value;
             this.Tags = Preconditions.CheckNotNull(tags, nameof(tags));
+
+            this.MetricKey = new Lazy<int>(() => this.GetMetricKey());
+        }
+
+        public bool Equals(Metric other)
+        {
+            return this.TimeGeneratedUtc == other.TimeGeneratedUtc &&
+                this.Name == other.Name &&
+                this.Value == other.Value &&
+                GetOrderIndependentHash(this.Tags) == GetOrderIndependentHash(other.Tags);
         }
 
         /// <summary>
         /// This combines the hashes of name and tags to make it easy to group all metrics.
         /// </summary>
         /// <returns>Hash of name and tags.</returns>
-        public int GetMetricKey()
+        int GetMetricKey()
         {
-            // TODO: replace with "return HashCode.Combine(Name.GetHashCode(), Tags.GetHashCode());"
-            // when upgraded to .net standard 2.1: https://docs.microsoft.com/en-us/dotnet/api/system.hashcode.combine?view=netstandard-2.1
-            int hash = 17;
-            hash = hash * 31 + this.Name.GetHashCode();
-            hash = hash * 31 + this.Tags.GetHashCode();
+            return CombineHash(this.Name.GetHashCode(), GetOrderIndependentHash(this.Tags));
+        }
 
-            return hash;
+        static int GetOrderIndependentHash<T1, T2>(IEnumerable<KeyValuePair<T1, T2>> dictionary)
+        {
+            return CombineHash(dictionary.Select(o => CombineHash(o.Key.GetHashCode(), o.Value.GetHashCode())).OrderBy(h => h).ToArray());
         }
 
-        public bool Equals(Metric other)
+        // TODO: replace with "return HashCode.Combine();"
+        // when upgraded to .net standard 2.1: https://docs.microsoft.com/en-us/dotnet/api/system.hashcode.combine?view=netstandard-2.1
+        static int CombineHash(int hash1, int hash2)
```

Comment:
```
        static int CombineHash(int hash1, int hash2)
        {
            return (17 * 31 + hash1) * 31 + hash2;
        }
```
Could be removed entirely since the second one uses params. I just made it b/c its slightly more efficient. 

Should I keep it for the efficiency or remove it for simplicity?

Reply:
Looks like you removed it :)